From c77dbf95dcd24f5ee9855f43a8dc3227b7b111b8 Mon Sep 17 00:00:00 2001
From: Konrad Leszczynski <konrad.leszczynski@intel.com>
Date: Tue, 15 Dec 2015 12:05:34 +0000
Subject: [PATCH 05/12] BACKPORT: usb: gadget: f_midi: use usb_gstrings_attach

In order to add configfs support the usb_gstrings_attach
must be used.

Change-Id: Ic74549e2a5ab371ac42071c4e7ea64ab40cdb44b
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-3355
Signed-off-by: Andrzej Pietrasiewicz <andrzej.p@samsung.com>
Signed-off-by: Felipe Balbi <balbi@ti.com>
Signed-off-by: Konrad Leszczynski <konrad.leszczynski@intel.com>
Reviewed-on: https://android.intel.com:443/448832
---
 drivers/usb/gadget/f_midi.c | 13 +++++++------
 1 file changed, 7 insertions(+), 6 deletions(-)

diff --git a/drivers/usb/gadget/f_midi.c b/drivers/usb/gadget/f_midi.c
index c40432b..7c075a9 100644
--- a/drivers/usb/gadget/f_midi.c
+++ b/drivers/usb/gadget/f_midi.c
@@ -718,6 +718,7 @@ static int f_midi_bind(struct usb_configuration *c, struct usb_function *f)
 	struct usb_midi_out_jack_descriptor_1 jack_out_emb_desc[MAX_PORTS];
 	struct usb_composite_dev *cdev = c->cdev;
 	struct f_midi *midi = func_to_midi(f);
+	struct usb_string *us;
 	int status, n, jack = 1, i = 0;
 
 	midi->gadget = cdev->gadget;
@@ -727,12 +728,13 @@ static int f_midi_bind(struct usb_configuration *c, struct usb_function *f)
 		goto fail_register;
 
 	/* maybe allocate device-global string ID */
-	if (midi_string_defs[0].id == 0) {
-		status = usb_string_id(c->cdev);
-		if (status < 0)
-			goto fail;
-		midi_string_defs[0].id = status;
+	us = usb_gstrings_attach(c->cdev, midi_strings,
+				 ARRAY_SIZE(midi_string_defs));
+	if (IS_ERR(us)) {
+		status = PTR_ERR(us);
+		goto fail;
 	}
+	ac_interface_desc.iInterface = us[STRING_FUNC_IDX].id;
 
 	/* We have two interfaces, AudioControl and MIDIStreaming */
 	status = usb_interface_id(c, f);
@@ -992,7 +994,6 @@ struct usb_function *f_midi_alloc(struct usb_function_instance *fi)
 	midi->qlen = opts->qlen;
 
 	midi->func.name		= "gmidi function";
-	midi->func.strings	= midi_strings;
 	midi->func.bind		= f_midi_bind;
 	midi->func.unbind	= f_midi_unbind;
 	midi->func.set_alt	= f_midi_set_alt;
-- 
1.9.1

